[WIP] Use weak aliases for the allocator shim

bjorn3 opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/142366)*
<!-- homu-ignore:end -->

jdonszelmann committed:
add aliasses to llvm

rust-log-analyzer commented:
The job **`mingw-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15586196007/job/43893040289) [(plain)](https://github.com/rust-lang/rust/commit/efcccbb478753849322234fd05871824625dee8d/checks/43893040289/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_codegen_gcc v0.1.0 (/checkout/compiler/rustc_codegen_gcc)
error[E0046]: not all trait items implemented, missing: `weak_alias`
  --> compiler/rustc_codegen_gcc/src/mono_item.rs:16:1
   |
16 | impl<'gcc, 'tcx> PreDefineCodegenMethods<'tcx> for CodegenCx<'gcc, 'tcx> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing `weak_alias` in implementation
   |
   = help: implement the missing item: `fn weak_alias(&self, _: <Self as BackendTypes>::Function, _: &str, _: &str) { todo!() }`

For more information about this error, try `rustc --explain E0046`.
[RUSTC-TIMING] rustc_codegen_gcc test:false 3.557
error: could not compile `rustc_codegen_gcc` (lib) due to 1 previous error
Build completed unsuccessfully in 0:02:43
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #142259) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bjorn3 committed:
[WIP] Use weak aliases for the allocator shim

bjorn3 commented:
Cleaned up a bit. Let's see if it didn't regress anything.

@bors2 try

bjorn3 committed:
Use the .drectve section for exporting symbols from dlls on Windows

While it would be reasonable to expect the Windows linker to handle
linker args in the .drectve section identical to cli arguments, as it
turns out exporting weak symbols only works when the /EXPORT is in the
.drectve section, not when it is a linker argument or when a .DEF file
is used.

bjorn3 commented:
Rebased on top of https://github.com/rust-lang/rust/pull/142568 which should have fixed the last remaining MSVC issue. MinGW is still broken because of ld.bfd having broken weak symbol handling.

@bors2 try jobs=x86_64-msvc-1,x86_64-msvc-2

bjorn3 commented:
Actually, maybe https://github.com/rust-lang/rust/pull/134522 is the better approach after all. It doesn't need an extra attribute for specifying the symbol aliasing and as it turns out weak definitions are implemented by LLVM as weak symbol aliases on Windows anyway.

@bors2 try cancel

bjorn3 commented:
Closing in favor of the EII work.
